arangod/Cluster/SynchronizeShard.cpp

```diff
@@ -384,89 +384,117 @@ static inline bool isStopping() {
   }
-
+  
   VPackBuilder body;
-  {
-    VPackObjectBuilder o(&body);
+  { VPackObjectBuilder o(&body);
     body.add(ID, VPackValue(std::to_string(rlid)));
     body.add(COLLECTION, VPackValue(collection));
     body.add(TTL, VPackValue(timeout));
-    body.add(StaticStrings::ReplicationSoftLockOnly, VPackValue(soft));
-  }
+    body.add(StaticStrings::ReplicationSoftLockOnly, VPackValue(soft)); }
 
   auto url = DB + database + REPL_HOLD_READ_LOCK;
-
-  cc->asyncRequest(clientId, 2, endpoint, rest::RequestType::POST, url,
-                   std::make_shared<std::string>(body.toJson()),
-                   std::unordered_map<std::string, std::string>(),
-                   std::make_shared<SynchronizeShardCallback>(this), timeout, true, timeout);
-
+  
+  auto const transactionId = TRI_NewTickServer();
+
+  // We deviate from 3.5 implementation, as we would like to be able to handle
+  // 3.3 servers, where the POST request is kept open until the lock is deleted.
+  
+  auto postres = cc->asyncRequest(
+    "", transactionId, endpoint, rest::RequestType::POST, url,
+    std::make_shared<std::string>(body.toJson()),
+    std::unordered_map<std::string, std::string>(),
+    std::make_shared<SynchronizeShardCallback>(this), timeout, true, timeout);
+  
   // Intentionally do not look at the outcome, even in case of an error
   // we must make sure that the read lock on the leader is not active!
-  // This is done automatically below.
+  // This is done automatically below. 
 
-  double sleepTime = 0.5;
-  size_t count = 0;
-  size_t maxTries = static_cast<size_t>(std::floor(600.0 / sleepTime));
-  while (++count < maxTries) {  // wait for some time until read lock established:
+  while (true) {  // wait for some time until read lock established:
 
+    // Exit immediately on shutdown 
     if (isStopping()) {
       return arangodb::Result(TRI_ERROR_SHUTTING_DOWN);
     }
 
-    // Now check that we hold the read lock:
-    auto putres = cc->syncRequest(clientId, 1, endpoint, rest::RequestType::PUT,
-                                  url, body.toJson(),
-                                  std::unordered_map<std::string, std::string>(), timeout);
+    // Timeout. Just to make sure that we are not holding the lock accidentally,
+    // we cannot exit the function here.
+    if (duration_cast<seconds>(steady_clock::now() - start).count() > timeout) {
+      break;
+    }
 
-    auto result = putres->result;
-    if (result != nullptr && result->getHttpReturnCode() == 200) {
-      auto const vp = putres->result->getBodyVelocyPack();
-      auto const& slice = vp->slice();
-      TRI_ASSERT(slice.isObject());
-      if (slice.hasKey("lockHeld") && slice.get("lockHeld").isBoolean() &&
-          slice.get("lockHeld").getBool()) {
+    // See if above POST has been returned (>= 3.4 only)
+    auto enqres = cc->enquire(postres);
+    if (enqres.status == CL_COMM_SENT) {
+      if (enqres.result != nullptr && enqres.result->getHttpReturnCode() == 200) {
```

Comment:
Need to call `cc->drop(postres)` in this case, otherwise the structure remains indefinitely in `ClusterComm` data structures.